kernel/kernel-api/src/main/java/io/delta/kernel/internal/skipping/DataSkippingUtils.java

```diff
@@ -0,0 +1,290 @@
+
+            // TODO more expressions
+            default:
+                return Optional.empty();
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////////////////////
+    // Useful helper functions for dealing with expressions
+    //////////////////////////////////////////////////////////////////////////////////
+
+    /** Construct the skipping predicate for a given comparator */
+    private static Predicate constructComparatorDataFilters(
+            String comparator, Column leftCol, Literal rightLit, StatsSchemaHelper schemaHelper) {
+
+        switch (comparator.toUpperCase(Locale.ROOT)) {
+
+            // Match any file whose min/max range contains the requested point.
+            case "=":
+                return new And(
+                    new Predicate("<=", schemaHelper.getMinColumn(leftCol), rightLit),
+                    new Predicate("<=", rightLit, schemaHelper.getMaxColumn(leftCol))
+                );
+
+            // Match any file whose min is less than the requested upper bound.
+            case "<":
+                return new Predicate("<", schemaHelper.getMinColumn(leftCol), rightLit);
+
+            // Match any file whose min is less than or equal to the requested upper bound
+            case "<=":
+                return new Predicate("<=", schemaHelper.getMinColumn(leftCol), rightLit);
+
+            // Match any file whose max is larger than the requested lower bound.
+            case ">":
+                return new Predicate(">", schemaHelper.getMaxColumn(leftCol), rightLit);
+
+            // Match any file whose max is larger than or equal to the requested lower bound.
+            case ">=":
+                return new Predicate(">=", schemaHelper.getMaxColumn(leftCol), rightLit);
+
+            default:
+                throw new IllegalArgumentException(
+                    String.format("Unsupported comparator expression %s", comparator));
+        }
+    }
+
+    private static Map<String, String> reverseComparators = new HashMap<String, String>(){{
+        put("=", "=");
+        put("<", ">");
+        put("<=", ">=");
+        put(">", "<");
+        put(">=", "<=");
+    }};
+
+    private static Predicate reverseComparatorFilter(Predicate predicate) {
+        return new Predicate(
+            reverseComparators.get(predicate.getName().toUpperCase(Locale.ROOT)),
+            getRight(predicate),
+            getLeft(predicate)
+        );
+    }
+
+    /**
+     * Return an expression cast as a predicate, throw an error if it is not a predicate
+     */
+    private static Predicate asPredicate(Expression expression) {
+        checkArgument(expression instanceof Predicate,
+            String.format("Expected predicate but got %s", expression));
+        return (Predicate) expression;
+    }
+
+    /**
+     * Utility method to return the left child of the binary input expression
+     */
+    static Expression getLeft(Expression expression) {
```

Comment:
Remind me why we don't just have a BinaryExpression class that supports a `.getLeft()` call?

Reply:
Main reason was to avoid the class bloat and keep the list of expression classes to a very minimal. We have an utility method in default handler, may be we should move that to kernel internal.

Reply:
hm. so `DataSkippingUtils` provides the expression get left / get right logic? Not, ExpressionUtils? This isn't cohesive.

See, this doesn't seem very findable. It's tough to search in the code for all helper/static/utility methods that take an Expression as input. ... espcially when some are in DataSkipppingUtils and some are in ExpressionUtils

Reply:
this seems like a very `c`, `struct` way of programming, not OOP. we should do a more OOP way. thoughts?

Reply:
the problem is we can use class/trait structure for this. That would have been the most OOP way. In absense of that, yeah, ExpressionUtils seems better.

Reply:
I can move these to `ExpressionUtils` in kernel-api and (and try to combine with the ones in kernel-defaults)

Reply:
Moved to ExpressionUtils